diff --git a/setup.py b/setup.py
index 96563c1a..80e6ae5d 100644
--- a/setup.py
+++ b/setup.py
@@ -122,10 +122,14 @@ def remove_output(*sources: str) -> Generator[None, None, None]:
 with remove_output("pybind11/include", "pybind11/share"):
     # Generate the files if they are not present.
     with TemporaryDirectory() as tmpdir:
-        cmd = ["cmake", "-S", ".", "-B", tmpdir] + [
+        cmd = ["cmake", '-GVisual Studio 16 2019', "-S", ".", "-B", tmpdir] + [
             "-DCMAKE_INSTALL_PREFIX=pybind11",
             "-DBUILD_TESTING=OFF",
             "-DPYBIND11_NOPYTHON=ON",
+            '-DPYTHON_EXECUTABLE:FILEPATH='+os.environ.get("PYTHON",""),
+            '-DPYTHON_LIBRARY:FILEPATH='+os.environ.get("PYTHON_LIBRARY",""),
+            '-DPYTHON_INCLUDE_DIR:PATH='+os.environ.get("PREFIX","")+ '\include',
+            '-DCMAKE_PREFIX_PATH='+os.environ.get("LIBRARY_PREFIX",""),
             "-Dprefix_for_pc_file=${pcfiledir}/../../",
         ]
         if "CMAKE_ARGS" in os.environ:

